lib/ansible/galaxy/dependency_resolution/dataclasses.py

```diff
@@ -584,3 +584,13 @@ def __new__(cls, *args, **kwargs):
 
     def __init__(self, *args, **kwargs):
         super(Candidate, self).__init__()
+
+    def with_signatures_repopulated(self):  # type: (Candidate) -> Candidate
+        """Populate a new Candidate instance with Galaxy signatures.
+        :raises AssertionError: If the supplied candidate is not sourced from a Galaxy-like index.
+        """
+        if self.type != 'galaxy':
+            raise AssertionError(f"Invalid collection type for {self}: unable to get signatures from a galaxy server.")
```

Comment:
@s-hertel in https://github.com/ansible/ansible/pull/80334#discussion_r1157395171 you implied that you wanted to use `repr()` for `self`, is that correct?
When you inject a variable into an f-string, it uses `str()`: `f'{var}'` becomes `f'{var !s}'` implicitly. To make use of `repr()`, use another specifier — `f'{var !r}'` .
```suggestion
            raise AssertionError(f"Invalid collection type for {self !r}: unable to get signatures from a galaxy server.")
```

Reply:
I also noticed that you didn't use `AnsibleAssertionError` but didn't understand why. https://docs.ansible.com/ansible/latest/dev_guide/testing/sanity/no-assert.html recommends using it whenever possible:
> [..] When possible, utilize a more specific exception than `AnsibleAssertionError`.

Did you just want to avoid importing this exception? What's the motivation here?

```suggestion
            raise AnsibleAssertionError(
                f"Invalid collection type for {self !r}: "
                "unable to get signatures from a galaxy server.",
            )
```

Reply:
Per https://github.com/ansible/ansible/blob/c1e19e4bddd13425aba45733372f2a676506256c/lib/ansible/errors/__init__.py#LL237C7-L237C27 `AnsibleInternalError` could be more precise, though...

Reply:
Because I had no idea there was a convention, or that there was documention for it, and I remembered you raising  assertion errors in your dependency resolver PR.

Reply:
Ah.. Yeah. Bare `AssertionError` is fine, I just remembered that this sanity check recommends an internal one. I'm fine with stdlib personally.

Reply:
@s-hertel also, if you didn't intend to use `!r`, feel free to mark this thread as resolved. I don't have a strong preference, just wanted to ask about the inconsistencies I noticed...

Reply:
Also, the repr issue was because I forgot we define `__str__` in addition to `__repr__`.

Reply:
Good thing we have reviews :)

Reply:
This is the reasons I prefer explicit specifiers everywhere, so that it's obvious what's expected on the calling side without relying on some disconnected definitions the reader might not even know to look up...

Reply:
I'm a bit torn over adding it honestly since it's to prevent developer error in the future - would better to have tests... I'd guess there are other places that would be more helpful to use __repr__ than __str__. Why is the latter even defined?

Reply:
Well, `repr()` is for devs/REPL/tracebacks/etc. while `str()` is for user-facing output usually.

Reply:
`__str__` won't necessarily identify the requirement it came from though, I think both devs and users benefit more from the `__repr__` implementation.

Reply:
Not sure but we could define these things better, in a separate effort. Typically `repr()` would also show `id(self)` somewhere in its output, for example. Just need to double-check that none of the internals rely on `str()`. I can imagine that something in the resolver might.

FWIW I should've added the docstrings for clarity when implementing these initially, but probably forgot. Magic methods deserve them just as much as other public APIs...